### ISSUE TITLE
stack-buffer-overflow in toktx

### ISSUE THREAD

**Issue by gy741** [KhronosGroup/KTX-Software#39]
Hello,

Building an environment with ASan, memory corruption occurs when the following command is executed.

```
make toktx-tests 
```

Thanks.

```
=================================================================
==7846==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7fff575443b4 at pc 0x000000448086 bp 0x7fff575440b0 sp 0x7fff57543840
WRITE of size 21 at 0x7fff575443b4 thread T0
    #0 0x448085 in scanf_common(void*, int, bool, char const*, __va_list_tag*) (/home/karas/KTX-Software/build/cmake/linux/Debug/toktx+0x448085)
    #1 0x449246 in sscanf (/home/karas/KTX-Software/build/cmake/linux/Debug/toktx+0x449246)
    #2 0x4de438 in readPAM(_IO_FILE*, unsigned int&, unsigned int&, unsigned int&, unsigned int&, unsigned int&, unsigned char*&) /home/karas/KTX-Software/tools/toktx/image.cpp:408:12
    #3 0x4dcd04 in readNPBM(_IO_FILE*, unsigned int&, unsigned int&, unsigned int&, unsigned int&, unsigned int&, unsigned char*&) /home/karas/KTX-Software/tools/toktx/image.cpp:151:10
    #4 0x4dfa25 in main /home/karas/KTX-Software/tools/toktx/toktx.cpp:273:17
    #5 0x7f5e9362c82f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:291
    #6 0x4358f8 in _start (/home/karas/KTX-Software/build/cmake/linux/Debug/toktx+0x4358f8)

Address 0x7fff575443b4 is located in stack of thread T0 at offset 372 in frame
    #0 0x4de0ff in readPAM(_IO_FILE*, unsigned int&, unsigned int&, unsigned int&, unsigned int&, unsigned int&, unsigned char*&) /home/karas/KTX-Software/tools/toktx/image.cpp:383

  This frame has 4 object(s):
    [32, 287) 'line'
    [352, 372) 'tupleType' <== Memory access at offset 372 overflows this variable
    [416, 420) 'maxval'
    [432, 436) 'depth'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow ??:0 scanf_common(void*, int, bool, char const*, __va_list_tag*)
Shadow bytes around the buggy address:
  0x10006aea0820: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006aea0830: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006aea0840: 00 00 00 00 00 00 00 00 f1 f1 f1 f1 00 00 00 00
  0x10006aea0850: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006aea0860: 00 00 00 00 00 00 00 00 00 00 00 07 f2 f2 f2 f2
=>0x10006aea0870: f2 f2 f2 f2 00 00[04]f2 f2 f2 f2 f2 04 f2 04 f3
  0x10006aea0880: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006aea0890: 00 00 00 00 f1 f1 f1 f1 00 00 00 00 00 00 00 00
  0x10006aea08a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10006aea08b0: 00 00 00 00 00 00 00 07 f3 f3 f3 f3 f3 f3 f3 f3
  0x10006aea08c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==7846==ABORTING
```

**Comment by msc-** [comment]
What branch are you building? What is ASan?

**Comment by gy741** [comment]
I am using a master branch.

AddressSanitizer (ASan): ASAN is a tool similar to valgrind.

```
AddressSanitizer (ASan) is a fast memory error detector based on compiler instrumentation (LLVM).
It is fully usable for Chrome on Linux and Mac. There's a mostly-functional Windows port in progress too, and you can using SyzyASan.
Additional info on the tool itself is available at http://clang.llvm.org/docs/AddressSanitizer.html.
```

**Comment by msc-** [comment]
Fixed in commit a439f903e. Need to allow space for terminating NUL when using %s format in scanf. Thanks for the report.
